perf: Improve performance of scalar Y -> i functions

caspervdw opened the pull request:
This extends the work started in https://github.com/shapely/shapely/pull/2339 for the integer functions.
Please check the PR description there for a description of what happened here.

**Additional changes**

- The integer function not only have a GEOS function as `data`, but also an `errcode` and `missing_value`. I changed the definition of this into a new struct type `Y_i_func_data`, to make the type casting less awkward.

**Notes**

- `get_num_points` is not implemented as a scalar version; possibly we can put it into `LineString.__len__` and `LinearRing.__len__`
- `get_num_coordinates`  same, but this could be implemented on a new property `BaseGeometry.n_coords`. `get_num_points` seems the same but then maybe more optimal for linestring/linearring
- `get_dimensions` is not implemented, it is a bit weird one because it is implied by the type, but still, for the geometrycollection it makes sense to provide it. Maybe a `BaseGeometry.dimensionality` property.
- `get_srid` is not implemented, it could be an `srid` property (but then again, it is up to debate on whether we want to "support" srids at all)

I leave these up to later implementation- don't want to mix a refactor with new features.

jorisvandenbossche commented:
> I leave these up to later implementation- don't want to mix a refactor with new features.

Indeed, let's discuss those separately

caspervdw commented on src/geos_funcs_Y_i.c:
```diff
@@ -0,0 +1,286 @@
+
+typedef struct {
+  FuncGEOS_Y_i* func;
+  int errcode;        // Value that indicates an error from GEOS
```
Short version: I don't want to change to much in the code, because it is just not in scope of my optimization now. It seems like most functions are just plain attribute access, or even const return values depending on the geometry type, which doesn't really need any exception handling. But I didn't analyze them all so it could be that some functions return `-1` and set an error message.

I can exclude that situation for the two functions that were configured with an errcode of `0`:

- `GEOSGeom_getDimensions_r`, it may return one of `-1, 0, 1, 2`, neither of them meaning error. In fact, you can analyse the code in [GEOS](https://github.com/search?q=repo%3Alibgeos%2Fgeos+%22%3A%3AgetDimension%22+path%3A%2F%5Esrc%5C%2Fgeom%5C%2F%2F&type=code) and there is really no way an exception can be set. So an error code is just not there for that one.
- `GEOSGetSRID_r` is just a direct attribute access. Semantically, `0` means "no SRID". It is a pity that we picked `-1` for that actually in the Python wrapper.

So in fact, the errcode can be hardcoded to `-1`, simplifying this configuration definition. The `null_value` cannot. I will push that change.

caspervdw commented:
> I apparently had two pending review comments that I forgot to submit

Happens to me regularly. Thanks for making the review. I addressed your comments. SHould be good to go now.

sgillies reviewed (approved):
@caspervdw thank you!
